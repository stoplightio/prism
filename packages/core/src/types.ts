/**
 * The below few interfaces (until end comment) should be defined in the @stoplight/types repo
 */

export interface IDisposable {
  dispose: () => void;
}

export type Path = Array<string | number>;

export enum ValidationSeverity {
  ERROR = 'error',
  WARN = 'warn',
  INFO = 'info',
}
export interface IValidation {
  path: Path;
  name: string;
  summary: string;
  severity: ValidationSeverity;
  message: string;
}

// END

export interface IPrism<Resource, Input, Output, Config, LoadOpts> {
  process: (input: Input, config?: Config) => Promise<IPrismOutput<Input, Output>>;
  load: (opts?: LoadOpts) => Promise<void>;
  readonly resources: Resource[];
}

export interface IPrismConfig {
  mock?: boolean | object;
  security?: boolean | object;
  validate?: boolean | object;
}

export type PrismConfigFactory<C, I> = (input: I, defaultConfig?: PrismConfig<C, I>) => C;
export type PrismConfig<C, I> = C | PrismConfigFactory<C, I>;

export interface ILoader<Options, Resource> {
  load: (opts?: Options, defaultLoader?: ILoader<Options, Resource>) => Promise<Resource[]>;
}

export interface IFilesystemLoaderOpts {
  path?: string;
}

export interface IHttpLoaderOpts {
  url?: string;
}

export interface IRouter<Resource, Input, Config> {
  route: (
    opts: { resources: Resource[]; input: Input; config?: Config },
    defaultRouter?: IRouter<Resource, Input, Config>
  ) => Resource;
}

export interface IForwarder<Resource, Input, Config, Output> {
  forward: (
    opts: { resource?: Resource; input: IPrismInput<Input>; config?: Config },
    defaultForwarder?: IForwarder<Resource, Input, Config, Output>
  ) => Promise<Output>;
}

export interface IMocker<Resource, Input, Config, Output> {
  mock: (
    opts: Partial<IMockerOpts<Resource, Input, Config>>,
    defaultMocker?: IMocker<Resource, Input, Config, Output>
  ) => Promise<Output>;
}

export interface IMockerOpts<Resource, Input, Config> {
  resource: Resource;
  input: IPrismInput<Input>;
  config: Config;
}

export interface IValidator<Resource, Input, Config, Output> {
  validateInput?: (
    opts: { resource: Resource; input: Input; config?: Config },
    defaultValidator?: IValidator<Resource, Input, Config, Output>
  ) => Promise<IValidation[]>;
  validateOutput?: (
    opts: { resource: Resource; output?: Output; config?: Config },
    defaultValidator?: IValidator<Resource, Input, Config, Output>
  ) => Promise<IValidation[]>;
}

export interface IPrismComponents<Resource, Input, Output, Config, LoadOpts> {
  config: PrismConfig<Config, Input>;
  loader: ILoader<LoadOpts, Resource>;
  router: IRouter<Resource, Input, Config>;
  forwarder: IForwarder<Resource, Input, Config, Output>;
  mocker: IMocker<Resource, Input, Config, Output>;
  validator: IValidator<Resource, Input, Config, Output>;
}

export interface IPrismInput<I> {
  data: I;
  validations: {
    input: IValidation[];
  };
}

export interface IPrismOutput<I, O> {
  input?: I;
  output?: O;
  validations: {
    input: IValidation[];
    output: IValidation[];
  };
}
