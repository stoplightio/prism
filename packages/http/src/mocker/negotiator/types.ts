import { IHttpOperationConfig } from '@stoplight/prism-http';
import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}

export type NegotiationOptions = IHttpOperationConfig;

export type NegotiatePartialOptions = {
  code: string;
  dynamic: boolean;
  exampleKey?: string;
};
