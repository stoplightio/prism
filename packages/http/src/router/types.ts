import { IHttpOperationEx } from '../types';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperationEx;
  pathMatch: MatchType;
  methodMatch: MatchType;
  serverMatch?: Nullable<MatchType>;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
