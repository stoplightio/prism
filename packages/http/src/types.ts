import { IPrism, IPrismComponents, IPrismConfig } from '@stoplight/prism-core';
import { IHttpOperation, Omit } from '@stoplight/types';

export type TPrismHttpInstance<LoaderInput> = IPrism<
  IHttpOperation,
  IHttpRequest,
  IHttpResponse,
  IHttpConfig,
  LoaderInput
>;

export type TPrismHttpComponents<LoaderInput> = Partial<
  IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IHttpConfig, LoaderInput>
>;

// TODO: should be complete | and in the @stoplight/types repo
export type IHttpMethod = 'get' | 'put' | 'post' | 'delete' | 'options' | 'head' | 'patch' | 'trace'; // ... etc

export interface IHttpOperationConfig {
  mediaType?: string;
  code?: string;
  exampleKey?: string;
  dynamic: boolean;
}

export interface IHttpConfig extends IPrismConfig {
  mock: false | IHttpOperationConfig;

  validate?: {
    request?:
      | boolean
      | {
          hijack?: boolean;
          headers?: boolean;
          query?: boolean;
          body?: boolean;
        };

    response?:
      | boolean
      | {
          headers?: boolean;
          body?: boolean;
        };
  };
}

export interface IHttpNameValues {
  [name: string]: string | string[];
}

export interface IHttpNameValue {
  [name: string]: string;
}

export interface IHttpUrl {
  baseUrl?: string;
  path: string;
  query?: IHttpNameValues;
}

export interface IHttpRequest {
  method: IHttpMethod;
  url: IHttpUrl;
  headers?: IHttpNameValue;
  body?: any;
}

export interface IHttpResponse {
  statusCode: number;
  headers?: IHttpNameValue;
  body?: any;
}

export type ProblemJson = {
  type: string;
  title: string;
  status: number;
  detail: string;
};

export class ProblemJsonError extends Error {
  public static fromTemplate(template: Omit<ProblemJson, 'detail'>, detail?: string): ProblemJsonError {
    Error.captureStackTrace(this, ProblemJsonError);
    return new ProblemJsonError(
      `https://stoplight.io/prism/errors#${template.type}`,
      template.title,
      template.status,
      detail || '',
    );
  }

  public static fromPlainError(error: Error & { detail?: string; status?: number }): ProblemJson {
    return {
      type: error.name && error.name !== 'Error' ? error.name : 'https://stoplight.io/prism/errors#UNKNOWN',
      title: error.message,
      status: error.status || 500,
      detail: error.detail || '',
    };
  }

  constructor(readonly name: string, readonly message: string, readonly status: number, readonly detail: string) {
    super(message);
    Error.captureStackTrace(this, ProblemJsonError);
  }
}

export type PayloadGenerator = (f: unknown) => Promise<unknown>;
