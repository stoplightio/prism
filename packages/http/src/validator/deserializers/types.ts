import { HttpParamStyles } from '@stoplight/types';
import { IHttpNameValue, IHttpNameValues, JSONSchema } from '../../types';

export type deserializeFn<Parameters> = (
  name: string,
  parameters: Parameters,
  schema?: JSONSchema,
  explode?: boolean
) => unknown;

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: deserializeFn<Parameters>;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
