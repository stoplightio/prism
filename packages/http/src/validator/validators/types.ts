import { IPrismDiagnostic } from '@stoplight/prism-core/src/types';
import { JSONSchema } from 'http/src/types';

export interface IHttpValidator<Target, Specs> {
  validate(target: Target, specs: Specs[], mediaType?: string): IPrismDiagnostic[];
}

export interface ISchemaValidator<S = JSONSchema> {
  validate(content: any, schema: S): IPrismDiagnostic[];
  supports(mediaType: string): boolean;
}

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: JSONSchema) => IPrismDiagnostic[]) | undefined;
}
