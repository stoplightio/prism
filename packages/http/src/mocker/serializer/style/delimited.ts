import { partial } from 'lodash';

function serializeAndImplode(separator: string, name: string, value: Array<string | number | boolean>) {
  return (
    encodeURIComponent(name) +
    '=' +
    value
      .map(String)
      .map(encodeURIComponent)
      .join(separator)
  );
}

function serializeAndExplode(name: string, value: Array<string | number | boolean>) {
  return value
    .map(String)
    .map(v => `${encodeURIComponent(name)}=${encodeURIComponent(v)}`)
    .join('&');
}

export function serializeWithDelimitedStyle(
  separator: string,
  name: string,
  value: Array<string | number | boolean>,
  explode?: boolean
): string {
  return explode ? serializeAndExplode(name, value) : serializeAndImplode(separator, name, value);
}

export const serializeWithCommaDelimitedStyle = partial(serializeWithDelimitedStyle, ',');
export const serializeWithSpaceDelimitedStyle = partial(serializeWithDelimitedStyle, ' ');
export const serializeWithPipeDelimitedStyle = partial(serializeWithDelimitedStyle, '|');
