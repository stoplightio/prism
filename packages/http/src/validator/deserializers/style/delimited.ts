import { HttpParamStyles } from '@stoplight/types';
import { IHttpNameValues, JSONSchema } from '../../../types';
import { IHttpQueryParamStyleDeserializer } from '../types';

export class DelimitedStyleDeserializer implements IHttpQueryParamStyleDeserializer {
  constructor(
    private readonly separator: string,
    private readonly styleName:
      | HttpParamStyles.PipeDelimited
      | HttpParamStyles.SpaceDelimited
      | HttpParamStyles.CommaDelimited,
  ) {}
  public supports(style: HttpParamStyles) {
    return style === this.styleName;
  }

  public deserialize(name: string, parameters: IHttpNameValues, schema?: JSONSchema, explode?: boolean) {
    const type = schema ? schema.type : undefined;
    const values = parameters[name];

    if (type === 'array') {
      return explode ? this.deserializeImplodeArray(values) : this.deserializeArray(values);
    } else {
      throw new Error('Space/pipe/comma.. delimited style is only applicable to array parameter');
    }
  }

  private deserializeImplodeArray(value: string | string[]) {
    return Array.isArray(value) ? value : [value];
  }

  private deserializeArray(value: string | string[]) {
    if (Array.isArray(value)) {
      // last query param is taken into account
      value = value[value.length - 1];
    }

    return value ? value.split(this.separator) : '';
  }
}
