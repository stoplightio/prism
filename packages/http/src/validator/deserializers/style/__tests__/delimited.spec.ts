import { HttpParamStyles } from '@stoplight/types';

import { DelimitedStyleDeserializer } from '../delimited';

describe('DelimitedStyleDeserializer', () => {
  const delimitedStyleDeserializer = new DelimitedStyleDeserializer('|', HttpParamStyles.PipeDelimited);

  describe('supports()', () => {
    describe('style is supported', () => {
      it('returns true', () => {
        expect(delimitedStyleDeserializer.supports(HttpParamStyles.PipeDelimited)).toBe(true);
      });
    });

    describe('style is not supported', () => {
      it('returns false', () => {
        // Force compile to pass for purpose of test.
        // @ts-ignore
        expect(delimitedStyleDeserializer.supports('invalid')).toBe(false);
      });
    });
  });

  describe('deserialize()', () => {
    describe('schema type is not array', () => {
      it('throws exception', () => {
        expect(() =>
          delimitedStyleDeserializer.deserialize('key', {}, { type: 'string' }, false)
        ).toThrowErrorMatchingSnapshot();
      });
    });

    describe('schema type is array', () => {
      describe('explode is set', () => {
        describe('query param is an array', () => {
          it('returns unmodified query param', () => {
            expect(delimitedStyleDeserializer.deserialize('key', { key: ['a'] }, { type: 'array' }, true)).toEqual([
              'a',
            ]);
          });
        });

        describe('query param is a string', () => {
          it('returns value converted to array', () => {
            expect(delimitedStyleDeserializer.deserialize('key', { key: 'a' }, { type: 'array' }, true)).toEqual(['a']);
          });
        });
      });

      describe('explode is not set', () => {
        describe('query param is an array', () => {
          it('splits last query param array element', () => {
            expect(
              delimitedStyleDeserializer.deserialize('key', { key: ['a|b|c', 'd|e|f'] }, { type: 'array' }, false)
            ).toEqual(['d', 'e', 'f']);
          });
        });

        describe('query param is a string', () => {
          it('splits query param value', () => {
            expect(delimitedStyleDeserializer.deserialize('key', { key: 'a|b|c' }, { type: 'array' }, false)).toEqual([
              'a',
              'b',
              'c',
            ]);
          });
        });
      });
    });
  });
});
