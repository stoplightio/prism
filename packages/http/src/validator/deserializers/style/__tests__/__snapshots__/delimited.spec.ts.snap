// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelimitedStyleDeserializer deserialize() schema type is not array throws exception 1`] = `"Space/pipe/comma.. delimited style is only applicable to array parameter"`;
